### ISSUE TITLE
Disk usage expands dramatically during training, causes 'Disk may be full' error.

### ISSUE THREAD

**Issue by jrosebr1** [apple/turicreate#139]
I checked the Issues but I haven't found anyone else posting this error so I'm not sure if it's related to my environment, something I am doing wrong, or a bug in the actual library/toolset.

I created a 18GB image dataset (the size of the resulting SFrame) for object detection on an Amazon EC2 p2.xlarge instance. The EC2 instance has 128GB of disk space.

I then went to train the model. The only terminal output indicating that the images and annotations were being used; however, as I monitored disk space I was surprised to see that it was slowly starting to eat up any free space.

I had 80GB of free space before starting training. This 80GB was eaten up and the following error reported:

```$ python train.py
Using 'image' as feature column
Using 'annotations' as annotations column
Traceback (most recent call last):
  File "train.py", line 17, in <module>
    model = tc.object_detector.create(data, max_iterations=ITERS)
  File "/home/ubuntu/.virtualenvs/turicreate/local/lib/python2.7/site-packages/turicreate/toolkits/object_detector/object_detector.py", line 240, in create
    instances = (dataset.stack(annotations, new_column_name='_bbox', drop_na=True)
  File "/home/ubuntu/.virtualenvs/turicreate/local/lib/python2.7/site-packages/turicreate/data_structures/sframe.py", line 5117, in stack
    new_column_type, drop_na))
  File "/home/ubuntu/.virtualenvs/turicreate/local/lib/python2.7/site-packages/turicreate/cython/context.py", line 49, in __exit__
    raise exc_type(exc_value)
IOError: Fail to write. Disk may be full.: iostream error```

Again, I'm not sure if this is a problem with using Turicreate on EC2 or if this is a problem with the training process.

**Comment by jrosebr1** [comment]
@znation I increased my disk to 200GB and still ran into the same error. Do you have any suggestions that I could try? I can provide the code + dataset if you would like. Thank you!

**Comment by igiloh** [comment]
Hi @jrosebr1, thank you for the detailed report.

Yes, if you could provide the data and code you have used (as well as OS version and any additional information) that would be really helpful.
Also - if you could try to find a smaller subset of the same dataset that does work properly, and check the peak hard drive usage during model creation, that would also be great.

Can you quickly describe how did you load the data and created the sframe?
Did you happen to be aggregating many smaller sframes? As this process might be a bit space inefficient.
Can you try to call `data.materialize()` before creating the model, check if it finished correctly, and if it help with the creation error?

**Comment by jrosebr1** [comment]
Hey @igiloh, thanks for getting back.

I am running Ubuntu 16.04 on an Amazon EC2 instance. The actual instance is my [deep learning AMI](https://www.pyimagesearch.com/2017/09/20/pre-configured-amazon-aws-deep-learning-ami-with-python/) which is publicly available. I created a Python 2.7 virtual environment and installed `turicreate` into it. Here is the kernel version:

```$ uname -a
Linux ip-172-31-16-14 4.4.0-1044-aws #53-Ubuntu SMP Mon Dec 11 13:49:57 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux```

Along with my `pip freeze` output for the virtual environment:

```certifi==2017.11.5
chardet==3.0.4
coremltools==0.6.3
decorator==4.1.2
graphviz==0.8.1
idna==2.6
imutils==0.4.4
mxnet-cu80==0.11.0
numpy==1.13.3
olefile==0.44
pandas==0.21.1
Pillow==4.3.0
prettytable==0.7.2
progressbar2==3.34.3
protobuf==3.5.0.post1
python-dateutil==2.6.1
python-utils==2.2.0
pytz==2017.3
requests==2.18.4
scipy==1.0.0
six==1.10.0
turicreate==4.0
urllib3==1.22```

The dataset I am using is from dlib.net's public data:

http://dlib.net/files/data/dlib_front_and_rear_vehicles_v1.tar

The goal of this dataset is to train an object detector that can recognize the front and rear ends of cars. Images were annotated using dlib's `imglab` tool.

I parsed `imglab`'s resulting .xml files using the following code:

https://gist.github.com/jrosebr1/8dff34056420b819576560c86ff6dd00

The gist is that I initialized `tc.SFrameBuilder`, loop over the image annotations, parsed the bounding boxes, and then added the individual image + bounding box to the `SFrameBuilder `object. Once all annotations had been parsed I called `.close` to save the database.

And then trained using this code:

https://gist.github.com/jrosebr1/b06901a2cdda3f48691665084fcda60d

Initially I thought that the training/testing split might have been causing an error so I tried without it -- still no dice.

I'll get started training on a smaller subset and call `data.materialize()` as well and get back to you.

**Comment by jrosebr1** [comment]
I'm not sure what `data.materialize()` is supposed to output but I didn't get anything back:

```>>> import turicreate as tc
>>> INPUT_DATA = "../vehicles/vehicles.sframe"
>>> data = tc.SFrame(INPUT_DATA)
>>> data.materialize()
```

**Comment by jrosebr1** [comment]
Originally, I had 6,198 total images in my training set. Some of the training images are high resolution (> 2,000px along their largest dimension). I'm not sure if that is a problem or not, but again, I wanted to share.

As a test, I reduced the total number of images to 200 and then tried to train again. Running `df -m` so can slowly start to see my disk space being eaten up:

```/dev/xvda1        126999 13767    113217  11% /
/dev/xvda1        126999 35329     91655  28% /
/dev/xvda1        126999 46558     80425  37% /
/dev/xvda1        126999 104176     22808  83% /```

Note that the actual training of the network never starts:

```$ python train.py
Using 'image' as feature column
Using 'annotations' as annotations column
```
Which seems to imply that it's some sort of dataset issue.

Interestingly, once the script errors out and exits the disk space is automatically reclaimed.

FWIW, I was able to train an object detector on [CALTECH-101](http://www.vision.caltech.edu/Image_Datasets/Caltech101/) as well as subset of [LISA](http://cvrr.ucsd.edu/LISA/lisa-traffic-sign-dataset.html) (~4,000 images) that were more moderately sized (~1000px along their largest dimension); _but_I still ran into a problem where disk space was being eaten up -- there was just a point where training actually started and disk consumption stopped.

**Comment by igiloh** [comment]
@jrosebr1 thanks again for the extremely detailed report!!!

SFrames are lazily evaluated, so SFrame operations which you're performing aren't actually being calculated until the data is required. The `sf.materialize()` method forces that SFrame to perform all of the calculations in its graph, so all previous operations done on this SFrame are finalized.
`materialize` is usually useful if you've performed several pre-processing steps, and are about to iterate over the same SFrame (as in the case of several training epochs etc.).
Nevertheless, materializing is irrelevant in your case, since SFrames built by `SFrameBuilder` will always be materialized.

However - using `SFrameBuilder` might explain the HD explosion, as the SframeBuilder uses a very large cache, as described in https://github.com/apple/turicreate/issues/102.

While I'm downloading the entire dataset to try to reproduce the problem - I'll be really glad if you could please upload a small subset of about 10-15 images, with their respective CSVs (after you conversion from xml). I wanna try to load the data a bit differently, without looping over the CSVs and adding images one by one.

**Comment by jrosebr1** [comment]
@igiloh No problem! Here are the first images and the CSV file containing their bounding boxes:

https://s3-us-west-2.amazonaws.com/static.pyimagesearch.com/turicreate/data.zip

Thanks for looking into this! I'm excited to get it running and do more tests. I would love to include a Turicreate example in my book and at the very least cover it on the PyImageSearch blog.

**Comment by igiloh** [comment]
Hi @jrosebr1,

There seem to be a bug in the code [gist you've provided.](https://gist.github.com/jrosebr1/8dff34056420b819576560c86ff6dd00).
In line 62 you check if `filename != imagePath`, but you never set `filename` after the first line of the first CSV. This means that instead of ending up with one line per image file - you end up with a huge SFrame that has a single line per annotation. Not to mention that all the lines have the same image in their `image` column (the very first one in the db).
I don't know if this is exactly the same code you've been using - but if it is, that might explain the exploding disk size you're getting.

Regardless of this bug, I've written a short script to load your images without using `SFrameBuilder`. Please see [here](https://gist.github.com/igiloh/43ffbbbc216dd264ad42463c9992bfc6).
My only assumption was that your xml to csv converter would add a header line at the beginning of each csv file, with the column names 'imagePath, startX, startY, endX, endY, label` (It can be done also without that, let me know if you need the code).

Please let me know if any of these solutions (fixing the bug or trying out my code) now works for you.
It would be really great to even try to compare them in terms of run time and disk space, if you'd like to.

**Comment by jrosebr1** [comment]
Hey @igiloh, thanks so much for looking into this. Oof, I feel like an idiot now. I should have caught that the `filename` was never updated.

Thanks for providing your example script. I modified it slightly and used it to rebuild my dataset. It worked great and I had no problems. 

This time I was able to actually start training. There was some bloat in disk space use but not to the point where all the space was eaten up:

![screen shot 2017-12-31 at 3 10 20 pm](https://user-images.githubusercontent.com/759645/34464116-b60004f0-ee40-11e7-8920-7b6c7dac1be6.png)

As you can see from the below screenshot that training started:

![screen shot 2017-12-31 at 3 10 26 pm](https://user-images.githubusercontent.com/759645/34464117-c03c329a-ee40-11e7-9b18-4a976cde55db.png)

But unfortunately after the first ~9 stops it looks like trained stopped entirely. My console sat like that for ~45 minutes until I killed the script.

Looking at the output of `nvidia-smi` you can see that my GPU usage is 0%:

![screen shot 2017-12-31 at 3 10 41 pm](https://user-images.githubusercontent.com/759645/34464122-e2a41b2c-ee40-11e7-88b7-d9cc7d97b2a6.png)

Perhaps this is a symptom of the database?

**Comment by igiloh** [comment]
@jrosebr1, I'm glad to hear that things are finally working for you now.
Note that turicreate **doesn't** use GPUs by default, as many users may not have the relevant hardware.
In order to train on GPUs you need to both [install the relevant dependencies](https://github.com/apple/turicreate#gpu-support) and [explicitly set the number of GPUs](https://github.com/apple/turicreate/blob/master/userguide/image_classifier/advanced-usage.md#using-gpus).

It seems that the user guide chapter for object detection doesn't mention GPU support at all, which might explain the confusion. I'll open another issue to improve the documentation.

**Comment by jrosebr1** [comment]
Thank you for the clarification on explicitly setting the number of GPUs. However, I found a typo in the [explicitly set the number of GPUs](https://github.com/apple/turicreate/blob/master/userguide/image_classifier/advanced-usage.md#using-gpus) documentation:

``
turicreate.config.set_num_gpus(-11)
```
I believe it should actually be:

```
turicreate.config.set_num_gpus(-1)
```

As using `-11` will end in the following error:

```
turicreate.toolkits._main.ToolkitError: num_gpus must be between -1 and 9223372036854775807
```

Once I correctly set the number of GPUs on my EC2 instance it was the same problem as last time. The GPU was being utilized for the first few steps, then GPU utilization drops to zero, and it appears that the script stops training.

This was not the case with my other test datasets so I assume this may just be related to my dataset and perhaps even using Turicreate in the Amazon cloud. I use `mxnet` daily on this instance (without Turicreate) and I don't have the same issue. I'll keep playing around with it.

**Comment by igiloh** [comment]
Hi @jrosebr1,
What you're describing is peculiar. Are you sure there is no GPU nor CPU activity at all? The process just hangs?
I'll mark it as a bug, we'll try to reproduce it. If you can send your data preparation+ training script that would be great.
I'll also appreciate if you can try a smaller subset of your dataset and see if it still hangs. Any additional info about which cases work and which doesn't might be helpful.
Thanks!

**Comment by igiloh** [comment]
Oh, and thanks for noticing the user guide typo. I'll fix it right away.

**Comment by jrosebr1** [comment]
@igiloh 

> If you can send your data preparation+ training script that would be great.

I can certainly do that. Would you like me to tar up the raw images + SFrame dump as well? I'm happy to do so I just don't want to post a public link to the entire archive. If you would like to entire tarball to aid in reproducing the error please let me know -- but I would need a way to privately share the link.

**Comment by srikris** [comment]
@gustavla Can you take a look at this? The hang suggests to me that something isn't quite right with this setup.

**Comment by gustavla** [comment]
@jrosebr1 I'm sorry about these issues. Thanks so much for such a detailed account, that is really useful! I am setting up a script to reproduce this now thanks to all your scripts and the dataset. However, I do think I know what's going on already. As part of the a pre-processing step, the object detector takes all images in the SFrame and decodes them. This is to avoid the overhead of constantly decoding them. However, your images are large and some take up to 23 MB uncompressed. The way the object detector uses the SFrame is not optimal, since normally out-of-memory datasets are well handled. This is definitely something I want to see improved in the object detector.

As a work-around before this has addressed, you can try to scale down your images a bit first. They will be scaled down before they are passed into the neural network anyway, so it won't make a difference for the model performance. You can do something like this to get the size down:

```
sf['image'] = sf['image'].apply(lambda image: tc.image_analysis.resize(image, image.width//4, image.height//4, image.channels))
```

Let me know if that works as a temporary work-around.

**Comment by jrosebr1** [comment]
@gustavla Got it, that makes sense. I'll be sure to give that a try. Will scaling the images in the way you suggested affect the bounding boxes? Will have to manually scale them as well or will they be automatically scaled with the images?

**Comment by jrosebr1** [comment]
@gustavla I went back to my Python script used to build the SFrame dataset and applied the image scaling as you suggested:

```
final["image"] = final["image"].apply(
	lambda image: tc.image_analysis.resize(image,
		image.width // SCALE, image.height // SCALE, image.channels))
```
As well as scaling the bounding boxes:

```
data["x"] /= SCALE
data["y"] /= SCALE
data["width"] /= SCALE
data["height"] /= SCALE
```

`data` is then combined into `final` later on.

**I'm happy to report that this method worked!**

There was _zero_ disk bloat and the training started immediately.

A big thank you to you and @igiloh for helping out with this.

Let me know if there is anything else I can do to help as well.

**Comment by gustavla** [comment]
@jrosebr1 I'm glad to hear that and that you remembered to rescale the bounding boxes too. I will still work on finding a better solution, so I'll leave this issue open until it has been addressed.
